### ISSUE TITLE
Add encryption support

### ISSUE THREAD

**Issue by albsod** [SolidHal/PrawnOS#43]
It would be nice to be able to encrypt at least a part of the file system, e.g. the home directory, and prompt the user already during the install process.

I tried to set up **encryptfs** by following [this guide](https://www.howtoforge.com/tutorial/how-to-encrypt-directories-and-partitions-with-ecryptfs-on-debian/), but it doesn't work because `modprobe ecryptfs` fails (the module can't be found in `/lib/modules/4.17.2-gnu/`).

An alternative is cryptsetup + luks. Could that easily be set up so that just the kernel partition remains unencrypted?

I'd be happy to try out any good ideas that are suggested here.

**Comment by SolidHal** [comment]
@albsod I added the ecrypt fs kernel config option in the latest commit. Unfortunately it didn't make it into the release I uploaded earlier, but you can either build it yourself or wait for the next time I feel like the repo is in a stable enough state to make a new release. 

I will definitely be looking into what option we can have for encryption once I get a bit more polish on PrawnOS

**Comment by albsod** [comment]
@SolidHal Nice! I'll try this out after reinstalling the next time.

**Comment by SolidHal** [comment]
ecrypt fs is what chromeos used, so supporting it on this hardware is possible and efficient.

**Comment by albsod** [comment]
> A sensible default fstab is created now. Not sure if not having one would really cause problems for this system, but it's good to have.
> 
> The default just contains
> 
> ```
> /dev/mmcblk2p2 / ext4 defaults 0 1
> ```
> 
> What did you put in your fstab?

`/dev/mmcblk0p1 /media/albsod/SDCARD ext4 r 0 0`

When I add your line above just after mine, then the system boots fine and I can log in via lightdm, but there all of my files are still encrypted.

> 
> This
> 
> ```
> mount.ecryptfs[184]: Error initializing key module [/usr/lib/arm-linux-gnueabihf/ecryptfs/libecryptfs_key_mod_gpg.so]; rc = [-22]
> ``
> seems funky. Does that file exist? 
> ```

Yes, it exists...

I've now tried something else:
From a root terminal, I created a new (unencrypted) text file in /home/albsod, then I ran mount -t ecryptfs /home/albsod /tmp/albsod. 
I then tried to view the text file but got this error:

Valid eCryptfs headers not found in file header region or xattr region, inode 273249
cat: /path/to/file: input/output error
As soon as I unmount, I'm able to access the file again.

**Comment by robinde** [comment]
Have you tried getting dm_crypt to run on this @SolidHal ?
I'm not sure how it's supposed to work with the funky partition layout of the c201. On other armv7 boards, the kernel is in an unencrypted `/boot` partition, along with an `initramfs` necessary for decrypting the root partition.
I got as far modifying the `cmdline` for re-packaging the kernel, then using `mkinitcpio` to generate an `initramfs` based on a kernel compiled from rockmyy. I modified a `mkinitcpio` preset from ALARM for armv7 boards. But on boot it's left hanging because it can't find `/dev/mapper/cryptroot`
Any ideas, or things you've tried?

**Comment by SolidHal** [comment]
Sorry for my delayed response. I'm focusing all of my free time for this on fixing the suspend issue right now. Once the basic functions of the laptop are mostly bug free I'll be able to focus more on encryption and security in general. 

@albsod I'm not really sure what is going wrong then :/ I'm not familiar with ecryptfs at all so I'll have to try all of this myself to give any valuable information. We should definitely be able to get it to work though. 

@robinde I haven't at all, sorry. I think I recall some discussion among the chromiumOS devs about dmcrypt but they ended up using ecryptfs on this device so thats what I'll probably focus on supporting first unless dmcrypt (or something else entirely) offers some obvious advantages. I'm not very familiar with either software at this point to be honest. So I'll spend some time researching what would make most sense.

**Comment by JeremyRand** [comment]
> they ended up using ecryptfs on this device so thats what I'll probably focus on supporting first unless dmcrypt (or something else entirely) offers some obvious advantages

eCryptfs doesn't encrypt metadata (e.g. number of files, directory structure, file sizes, permissions, and similar stuff), while LUKS / dm-crypt does.  For this reason, LUKS / dm-crypt is substantially more secure than eCryptfs.

Also I got LUKS working (though I didn't yet integrate it into the install script), see https://github.com/SolidHal/PrawnOS/pull/65

**Comment by robinde** [comment]
Looking good @JeremyRand !
Have you tried some normal usage of your c201 with either your home or root partition encrypted?
With both ecryptfs or dm-crypt on arch, my c201 eventually crashes, feels a bit like the IO just gets jammed. Usually happens while browsing with firefox. You need to force a shutdown and some files tend to get corrupted, eventually needing a re-install to get it working smoothly again.

**Comment by JeremyRand** [comment]
@robinde After a few weeks of running PrawnOS with `/home` encrypted via LUKS, I can confirm that I do see some occasional filesystem corruption of `/home`.  I'm going to try to debug it more thoroughly when I have a few hours.  I did notice anecdotally that manually running `sync` from a terminal before shutting down seemed to significantly reduce the likelihood that I'd be asked to run `fsck` on the next boot, which makes me wonder whether there's some kind of kernel configuration or filesystem configuration bug that's causing the system to shut down before all the writes are properly flushed to storage.  I might try mounting the physical filesystem and/or the LUKS-contained filesystem with the `sync` flag to see if that helps.

**Comment by JeremyRand** [comment]
@robinde Can you confirm whether https://github.com/SolidHal/PrawnOS/pull/91 fixes the filesystem corruption that you were seeing with ecryptfs and dm-crypt on the C201?  Preliminary testing on my end indicates that it fixes LUKS, but I haven't let it run long enough to be certain.

**Comment by atopuzov** [comment]
Hi, I'm not using PrawnOS but I had similar issues on Arch and always had to disable `rk-crypto` as it would corrupt the filesystem. Disabling it fixes the issue.

**Comment by robinde** [comment]
@JeremyRand Nice find, let me try to build this again and see what it's like. I've been running parabola on my c201 with the 4.19 rockmyy kernel the last months.
How has your experience been? Using an encrypted home partition or encrypted root?

**Comment by SolidHal** [comment]
@robinde Ive made some improvements to the build system so that the filesystem only needs to be built once (since that is the long part) so make sure to check out the readme. Please let me know if any part of it isn't clear.

**Comment by ifbizo** [comment]
I'm also working on LUKS encryption with encrypted root. Currently running into two separate initramfs issues:

1. Builtin keyboard doesn't work
I tried adding cros_ec_spi and other related modules into initramfs and also tried building them direct into kernel, but still not working.
2. Missing /sbin/cryptsetup binary in initramfs
I see the binary exists inside the initramfs if I uncompress it manually, but it seemingly "disappears" during boot. I tried increasing kernel configs for initramfs size but no luck. Seems similar to https://stackoverflow.com/questions/31524636/files-disappearing-from-initramfs

**Comment by ifbizo** [comment]
I'll submit a pull request soon, I have it working on my tree: https://github.com/ifbizo/PrawnOS/tree/luks 
1. I haven't enabled the install to MMC yet
2. Need to change the password prompting

**Comment by sokolgeo** [comment]
Meanwhile, with latest release of PrawnOS, I followed https://wiki.debian.org/TransparentEncryptionForHomeFolder 

Worked fine, my home directory seem to be encrypted just fine. 

Hope it helps and thank you for the great work on PrawnOS!

**Comment by SolidHal** [comment]
@sokolgeo There have been reports of user data corruption, so I would watch out for that.

> Hope it helps and thank you for the great work on PrawnOS!
Thanks!

**Comment by sokolgeo** [comment]
@SolidHal I confirm data corruption, sadly. 
Now trying to build `sudo PRAWNOS_LUKS=yes make image` from https://github.com/ifbizo/PrawnOS/tree/luks
Encryption is critical to have for any real work with mobile device which could be lost or stolen.

**Comment by sokolgeo** [comment]
Building https://github.com/ifbizo/PrawnOS/tree/luks fails for me.

* Specifically, building with `sudo PRAWNOS_SUITE=buster PRAWNOS_LUKS=yes make image` results in image, which stucks at boot, apparently unable to mount encrypted volume it has just itself produced during the build process. 

* Building `sudo PRAWNOS_SUITE=stretch PRAWNOS_LUKS=yes make image` fails with error during the build time.

I am out of my options, unfortunately. Being nice idea, for me the PrawnOS is not safe to use in real life environment, where my device could get stolen, with lots of passwords and private data on it, unencrypted (!)

Please, any idea of getting at least '/home' directory reliably encrypted in PrawnOS, anyone?

**Comment by SolidHal** [comment]
>  Encryption is critical to have for any real work with mobile device which could be lost or stolen.
@sokolgeo I completely agree. Unfortunately, #102 is pretty tricky and for many people functioning internet is more important so I've had to dump most of my debug time into that. 

There are some patches in kernel 5.2 for the c201 related to encryption so it'll be worth testing encryption again when I get 5.2 functional.

**Comment by SolidHal** [comment]
Beginning to work actively on this issue. 
Looking into which method of encryption we should use, I found this https://www.phoronix.com/scan.php?page=article&item=ext4-crypto-418&num=1 which benchmarks ecryptfs, ext4 encryption, and dmcrypt/luks. Obviously this is a completely different system than the c201, but it is a good place to start. 

Based on that, I'll begin with LUKS support and test out some benchmarks vs a non-encrypted system.

**Comment by sokolgeo** [comment]
Thank you @SolidHal for great news! Would be glad to help you with testing these PrawnOS images!

**Comment by riwale** [comment]
LUKS version 1 works very good on latest release, will test v2, but expect similar results. Thanks so much, only reason for me to keep another chipset around.

**Comment by SolidHal** [comment]
@riwale What process are you going through to setup luks?

**Comment by riwale** [comment]
i downgraded to luks version 1 to use an encrypted volume through openwrt --> sshfs on a prawnos c201p. This image on a stick mounted gracefully on the book now.

i'm looking into encrypting my home folders, as they are mounted on the sd card, will try luks v2 on those.

Still - as encryption now works this way or another this book, wiith your work is clearly the go to choice for anyone looking for complete freedom and security.

**Comment by ifbizo** [comment]
> Building https://github.com/ifbizo/PrawnOS/tree/luks fails for me.
>     * Building `sudo PRAWNOS_SUITE=stretch PRAWNOS_LUKS=yes make image` fails with error during the build time.
> 
stretch is insignificant now, but curious what was the build error?

I'm using the same method as https://github.com/ifbizo/PrawnOS/commit/c29a4d9a3bbee2e828410cd776fb792517ea1b12 and I'm booting buster with kernel 4.19.71. Encrypted root, unencrypted boot partition.

I tried merging the 5.2 changes but don't get the kernel to boot. Will keep trying this week. I am using LUKS 1 as @riwale was. I couldn't get LUKS 2 to work, something with the versions used in Debian is causing problems.

**Comment by sokolgeo** [comment]
@SolidHal - sorry, could not replicate build error with stretch. 
Today the same virtual machine I used last time gets the build done OK. Go figure.

**Comment by ifbizo** [comment]
I created a new branch, https://github.com/ifbizo/PrawnOS/tree/luks2 which is a clean merge of my previous code that incorporates the latest PrawnOS commits. It's working with LUKS v2 on kernel 5.2.9, but without the hardware acceleration. I get actual kernel crashes with the kernel config ``CRYPTO_DEV_ROCKCHIP`` enabled (for 4.19 I'd only get corruption), but haven't debugged further. I can reproduce it by running ``cryptsetup benchmark``. FWIW, my c201 gets ``~92.9 MiB/s`` for ``aes-xts/256b`` without hardware acceleration.

@SolidHal if you want I can create a pull request, but I still see some work remaining:
- Password is requested twice during the build, once initially and again to reopen towards the end
- No install to MMC script yet
- No configuration of cryptsetup parameters (and these might not be set to good defaults)

If anyone has any other suggestions or changes, let me know!

**Comment by SolidHal** [comment]
@sokolgeo weird! Well, I'm glad it working now atleast.

@ifbizo Yeah, sorry, I spent many days trying to get the old config to work properly with 5.2 and never found the option that made it fail to boot. So I whipped up a fresh config for 5.2. I should have mentioned that somewhere. 


> It's working with LUKS v2 on kernel 5.2.9, but without the hardware acceleration. I get actual kernel crashes with the kernel config CRYPTO_DEV_ROCKCHIP enabled (for 4.19 I'd only get corruption), but haven't debugged further. I can reproduce it by running cryptsetup benchmark. FWIW, my c201 gets ~92.9 MiB/s for aes-xts/256b without hardware acceleration.

Hmm, thats no good. I'll pull a more recent version of 5.2 when I get a chance so we can test that. Maybe I'll have to dig up some more patches. Thanks for the testing information.

> @SolidHal if you want I can create a pull request, but I still see some work remaining:
> 
>     Password is requested twice during the build, once initially and again to reopen towards the end
>     No install to MMC script yet
>     No configuration of cryptsetup parameters (and these might not be set to good defaults)
> 
> If anyone has any other suggestions or changes, let me know!
> 

Optimally, I'd like to be able to only distribute one image with or without root encryption and have the setup occur in `installtoInternal.sh`. For booting off of the internal emmc I am most of the way through a system to enable this. Just finishing up the  custom initramfs. 
Of course, this would be impossible to achieve for those who want to boot off of an external usb or sd card since root encryption needs to be done before the filesystem is written. 
So, I'd be happy to take a pull request with your changes to allow people that specifically want to boot from an encrypted external device to build a custom image just for themselves.

One other thought I just had was we could provide the same system I'm implementing for the emmc in an "installToExternal.sh`. This would basically do everything `InstallToInternal.sh` does but instead allow the user to target an external device. This way, users that don't have the means to build their own custom encrypted image could still have the option to boot from an encrypted external device. 

Open to opinions here, not sure if people prefer to boot from external devices or the internal emmc.

**Comment by tehbra1n** [comment]
@SolidHal mostly as a matter of simplicity I've been booting off the internal emmc.  If there was a motivating reason to boot from external media it wouldn't be a huge deal.

**Comment by ifbizo** [comment]
@SolidHal I've also been booting off internal emmc. I'm just using external media for testing, which is why I enabled it there first. I agree with your optimal approach, though!

**Comment by JeremyRand** [comment]
> Open to opinions here, not sure if people prefer to boot from external devices or the internal emmc.

I typically boot PrawnOS from an external SD card, because I run Bitcoin Core on my C201, and the Bitcoin blockchain doesn't fit on the internal eMMC.  Also I like to be able to boot into ChromeOS for the rare cases where it's needed (although this is becoming less relevant now that PrawnOS has advanced so far).  So, I would definitely like the ability to install an encrypted PrawnOS to an external device.

**Comment by sokolgeo** [comment]
For my use case 16 GB internal memory is not enough, so I support focusing on _encrypted_ external SD card as primary storage option. 

On related topic - could we decouple volume encryption password from user unlock password? Otherwise one should set really long/complicated encryption password to prevent brute force password crack for data volume. But then one should input this complex password  on each screen unlock, which could be hassle. Any in- between proposals on having string data volume encryption password and relatively easy user screen unlock password? Thank you.

**Comment by rk-zero** [comment]
> For my use case 16 GB internal memory is not enough, so I support focusing on _encrypted_ external SD card as primary storage option. On related topic - could we decouple volume encryption password from user unlock password? Otherwise one should set really long/complicated encryption password to prevent brute force password crack for data volume. But then one should input this complex password on each screen unlock ...

Those two passwords are not related and can totally be different. User passwords are never the same as luks encrpyt, simply because there are more than one user on your device, even in default root and user.

**Comment by sokolgeo** [comment]
Thank you for clarification!

 Data corruption problem on encrypted volume hit me last time I tried enabling encryption with PrawnOS. Do you please know if data corruption on encrypted volume was still observed in recent PrawnOS builds?

**Comment by rk-zero** [comment]
The latest kernel patches regarding mmc were introduced 5 months ago, if your release is older than that you should try a more recent one.

Integrity itself is handled by the mmc, check health with the mmc-utils package, then run:

sudo mmc extcsd read /dev/mmcblk2 | grep LIFE
eMMC Life Time Estimation A [EXT_CSD_DEVICE_LIFE_TIME_EST_TYP_A]: 0x01
eMMC Life Time Estimation B [EXT_CSD_DEVICE_LIFE_TIME_EST_TYP_B]: 0x01

means 1-10% of lifetime used.

bad sectors:

sudo mmc extcsd read /dev/mmcblk2 | grep BAD
Bad Block Management mode [SEC_BAD_BLK_MGMNT]: 0x00

So zero.

Data corruption shouldn't be an issue right now.

**Comment by rk-zero** [comment]
I guess it's possible to bundle a script that skips the mmc.partmap and just creates a partition over the rest of the mmc / sd, i will try my luck on that. Maybe mmc0/2 will be an issue in unencrpytion during start and needs to be updated. Still totally doable, just needs some work.

**Comment by rk-zero** [comment]
@JeremyRand @sokolgeo 

I rewrote the script, there you go: https://github.com/rk-zero/PrawnOS/commit/e954e3a8653f5d3df91ad228553058b04aba5897

There are still minor bugs, debug outputs and i'm still learning, but it works so far, so good.

I successfully doubled a running install, even encrypted on an sd card unencrypted and doubled also on a usb drive with encryption so far.

**Comment by SolidHal** [comment]
Thorough testing is definitely still required to ensure no corruption occurs, but there’s been a handful of people using encryption for about a month and a half with no issues. 
Seems instability is still seen when using the Rockchip crypto driver, so we have that disabled for now. 

Regarding the size constraints of the mmc, I will work on getting support for encryption working on an external device. Looks like @rk-zero has a good start. 

That said, the setup I would suggest is to install PrawnOS to the emmc, and the using an as card as a an “extended” root file system. This is because the emmc has approximately double read/write speeds over the speeds the sd and usb devices are stuck at. So keeping the main OS files on the emmc will help you keep your patience with this device ;)
Maybe we can figure out a way to do this seamlessly.

**Comment by rk-zero** [comment]
This is a great idea, still the eMMC has so much issues and is only double the speed of the SD, even though 5-6 times faster than USB and lots more iops.

Can we talk about issues from the eMMC in another issue/channel? I saw you signed off some kernel patches considering this and i'm hitting roadblocks trying to actually zero out the device for correct partitioning.

Setting up a lvm and cache writes to the emmc makes sense, but the sd reader is not usb bottlenecked, you can get 50-60MBs out of it which seems fine for everyday use.

**Comment by SolidHal** [comment]
Closing, encryption is fully functional on the internal emmc and external devices (usb/sd card).

**Comment by sokolgeo** [comment]
Wow, thank you for implementing the encryption! 
Finally my Chromebook becomes useful!
